### ISSUE TITLE
Predict slow

### ISSUE THREAD

**Issue by bitnick10** [microsoft/onnxruntime#1126]
**Describe**
```
 auto start = std::chrono::steady_clock::now();
    for (int i = 0; i < 16000; i++) {
        OrtRun(session, NULL, input_node_names.data(), (const OrtValue * const*)&input_tensor, 1, output_node_names.data(), 1, &output_tensor);
    }
 auto end = std::chrono::steady_clock::now();
 std::cout << "elapsed:" << (double)std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() / 1000 << "s\n";
```
This takes 0.3s and same model using keras predict only takes 0.01s.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):Windows
- ONNX Runtime installed from (source or binary):nuget
- ONNX Runtime version:0.4
- Python version:
- Visual Studio version (if applicable):15.9.4

**Comment by bitnick10** [comment]
python code:
```
import keras
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
from keras.optimizers import SGD
import keras.engine.saving
import keras.models
import math
import numpy as np
import stopwatch
import onnxmltools
data_size = 16000
x_train = np.array(np.random.rand(data_size,39),dtype='float32')
y_train = np.zeros(shape=(data_size,2),dtype='float32')

model = Sequential()
model.add(Dense(units=22,input_shape=(39,), activation='tanh'))
model.add(Dense(units=22, activation='tanh'))
model.add(Dense(units=2, activation='tanh'))
sgd = SGD(lr=0.01, momentum=0.9)
model.compile(loss='categorical_crossentropy',optimizer=sgd,)

sw = stopwatch.Stopwatch()
sws=[]
for i in range(2):
    sw = stopwatch.Stopwatch()
    sw.reset();
    sw.stop();
    sws.append(sw)

for i in range(100):
    sws[0].start()
    model.train_on_batch(x_train, y_train)
    sws[0].stop()
    sws[1].start()
    y_predict=model.predict(x_train,batch_size=len(x_train))
    sws[1].stop()
    print("{0} {1:.3f} {2:.3f} ".format(i,sws[0].duration/(i+1),sws[1].duration/(i+1)))

onnx_model = onnxmltools.convert_keras(model, target_opset=7)
onnxmltools.save_model(onnx_model,"model.onnx")
```

**Comment by KeDengMS** [comment]
It is not an apple-to-apple comparison. In your Keras model, the batch size is 16000, while in ORT the batch size is 1, with 16000 iterations. 
@wenbingl can you check if the converter is able to generate model with batch dimension? This model has output with symbolic batch dim, but input batch size is 1...

As a workaround, you can manually edit model.onnx to add a symbolic batch dimension. My test shows ORT is 33% faster after the edit (0.6ms vs 0.9ms in Keras)
```
import onnx
mp = onnx.load_model('model.onnx')
mp.graph.input[0].type.tensor_type.shape.dim[0].dim_param = 'None'
onnx.save(mp, 'model_mod.onnx')

import onnxruntime
sess = onnxruntime.InferenceSession('model_mod.onnx')
import numpy as np
import time
data = np.random.rand(16000,39).astype(np.float32)
t = time.time()
results = sess.run([], {'dense_1_input_01':data})
t2 = time.time()
print(t2 - t)
```

**Comment by bitnick10** [comment]
So it is still the "batch" predict problem that I asked at #853 !

**Comment by KeDengMS** [comment]
If the ask is to use C api to edit model to have batch dimension, you could do that too, with ONNX protobuf directly. However that is beyond the scope of ONNXRuntime's C api..

**Comment by bitnick10** [comment]
I want to know how to predict like
 `y_predict=model.predict(x_train,batch_size=len(x_train)` in C or C++

**Comment by KeDengMS** [comment]
ONNXRuntime currently do not have that kind of C api, and batch dimension needs to be specified as symbolic instead of fixed size in the onnx model to allow batched predict.

I understand the usefulness of the proposed API, but there are ambiguities if the API is not carefully designed. In the case of 1 input, it's easy to replace 1 at dim0 with symbolic dimension, but we should also consider models with more than one inputs, and which dim of 1 in the inputs corresponds to batch is hard to guess.

For now, I think you need to manually edit the model to have symbolic batch dimension as the example code above. You may edit the model offline with Python, then run inference in C/C++ by feeding data with actual batch dimensions.

**Comment by bitnick10** [comment]
@KeDengMS Thanks for your patient answers.
